src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs

```diff
@@ -3,17 +3,129 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics.Contracts;
+using System.Runtime.CompilerServices;
 
 namespace System.Dynamic.Utils
 {
     internal static partial class CollectionExtensions
     {
+        public static T[] AddFirst<T>(this IList<T> list, T item)
+        {
+            T[] res = new T[list.Count + 1];
+            res[0] = item;
+            list.CopyTo(res, 1);
+            return res;
+        }
+
         public static T[] AddLast<T>(this IList<T> list, T item)
         {
             T[] res = new T[list.Count + 1];
             list.CopyTo(res, 0);
             res[list.Count] = item;
             return res;
         }
+
+        public static T[] RemoveFirst<T>(this T[] array)
+        {
+            T[] result = new T[array.Length - 1];
+            Array.Copy(array, 1, result, 0, result.Length);
+            return result;
+        }
+
+        public static T[] RemoveLast<T>(this T[] array)
+        {
+            T[] result = new T[array.Length - 1];
+            Array.Copy(array, 0, result, 0, result.Length);
+            return result;
+        }
+
+        // Name needs to be different so it doesn't conflict with Enumerable.Select
+        public static U[] Map<T, U>(this T[] array, Func<T, U> select)
+        {
+            int count = array.Length;
+
+            U[] result = new U[count];
+
+            for (int i = 0; i < count; i++)
+            {
+                result[i] = select(array[i]);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Wraps the provided enumerable into a ReadOnlyCollection{T}
+        ///
+        /// Copies all of the data into a new array, so the data can't be
+        /// changed after creation. The exception is if the enumerable is
+        /// already a ReadOnlyCollection{T}, in which case we just return it.
+        /// </summary>
+        [Pure]
```

Comment:
The is about the only type with this attribute used on it. Does it give any benefit with such limited use?